### PR TITLE
Backend plugins: Refactor to allow shared contract between core and external backend plugins

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
Refactor to allow shared contract between core and external backend plugins allowing core backend data sources in Grafana to be implemented in same way as an external backend plugin.

This is currently based on #25454, but when that is merged I will rebase this one.
Also dependent on https://github.com/grafana/grafana-plugin-sdk-go/pull/187

**Special notes for your reviewer**:
- Verified image renderer v2.0.0 (v2 protocol) works
- Verified image renderer v1.0.12 (v1 protocol) works
- Verified test data source works (v2 protocol)
- Verified adding a temporary new core data source for manual test/verification when calling resource (see `testdatasourcev2` package)

### CODE REVIEW

**Comment by torkelo** [comment]
Nice!!! ❤️

**Comment by kylebrandt** [comment]
<3 So much wrapping :exploding_head: :-)

**Review by bergquist — COMMENTED** [review]
:+1: 

Looking forward to see all those layers removed eventually :)
Great work!

How do we progress from here? Create a new PR based on this that updates all packages in `tsdb` ? This one is quite large already :)

**Review by kylebrandt — APPROVED** [review]
Very much appreciated :-)

File: .circleci/config.yml [hunk 1]
```diff
@@ -828,9 +828,11 @@ jobs:
               -E varcheck -E goconst -E errcheck -E staticcheck ./pkg/...
             ./scripts/go/bin/revive -formatter stylish -config ./scripts/go/configs/revive.toml ./pkg/...
             ./scripts/go/bin/revive -formatter stylish -config ./scripts/go/configs/revive-strict.toml \
+              -exclude ./pkg/plugins/backendplugin/pluginextensionv2/... \
               ./pkg/services/alerting/... \
               ./pkg/services/provisioning/datasources/... \
-              ./pkg/services/provisioning/dashboards/...
+              ./pkg/services/provisioning/dashboards/... \
+              ./pkg/plugins/backendplugin/...
```
Comment (bergquist) [inline .circleci/config.yml:835]: :clinking_glasses:

File: pkg/plugins/backendplugin/plugin.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+// Plugin backend plugin interface.
+type Plugin interface {
+	PluginID() string
```
Comment (bergquist) [inline pkg/plugins/backendplugin/plugin.go:12]: Would be great if these functions could have comments.

**Review by bergquist — APPROVED** [review]
:+1:

File: pkg/plugins/backendplugin/plugin.go (hunk 2 above)
Comment (marefr) [reply pkg/plugins/backendplugin/plugin.go:12]: Will consider in future PR.

**Pull request merged**